model/riscv_insts_zba.sail

```diff
@@ -71,8 +71,8 @@
 /* ****************************************************************** */
 union clause ast = RISCV_SLLIUW : (bits(6), regidx, regidx)
 
-mapping clause encdec = RISCV_SLLIUW(shamt, rs1, rd) if haveZba() & sizeof(xlen) == 64
-  <-> 0b000010 @ shamt @ rs1 @ 0b001 @ rd @ 0b0011011 if haveZba() & sizeof(xlen) == 64
+mapping clause encdec = RISCV_SLLIUW(shamt, rs1, rd) if (haveZba() | haveBExt()) & sizeof(xlen) == 64
```

Comment:
B implies Zba, so this is just an ISA parsing (which doesn't really exist) issue? This would be if Zba and B were independent extensions that both implement the same instructions (like M vs Zmmul).

Reply:
The proposed B extension include instructions provided by Zba, Zbb, and Zbs extensions.

Reply:
My reading of the spec is that B *is* Zba_Zbb_Zbs, not that it's a *separate* extension. And I don't see why you'd want it to be anything else?

Like how G is IMAFD; we just say "has F" in the Sail model, since G implies F, rather than putting "has F or G" everywhere.

Reply:
Yes include Zba, Zbb, Zbs so B=Zba+Zbb+Zbs. However, if B is not present i.e. misa.B is 0, then one or more of Zba, Zbb, or Zbs may or may not be present.

Reply:
Yes, I know. The point is, "has B" and "has B or Zba" are entirely equivalent, and it's the job of a centralised ISA parser to turn B into B_Zba_Zbb_Zbs, just like a toolchain does. Not the job of every single instruction to list every single extension that (transitively) provides that instruction. That would get very out of hand very quickly, and isn't helpful.

Reply:
So you would prefer if misa.B was set to 1 then haveZba be also set to 1?

Reply:
Yes, because that's what B actually means

Reply:
(Though note that haveZba is already hard-coded to true, so there's nothing to do there; in future the Sail model will have a proper ISA string parser, and then it'll have to include all the same rules toolchains have)

Reply:
But that not allow expressing behavior of clearing B clearly. For instance, if misa.B and haveZba was both 1 at reset. Then clearing B would leave Zba available. But if misa.B was 1 and haveZba was 0 at reset then clearing B would disable Zba. Forcing state of Zba to misa.B would not allow expressing this.

Reply:
I like the idea that @jrtc27 brought up to defer this to when we have isa string parsing implemented in sail. I will revise this PR to only add the misa.B and if misa.B is 1 set Zba/Zbb/Zbs as supported. For now clearing misa.B will keep Zba/Zbb/Zbs as still available and that option for them to go away can be done when there is ISA string support.

Reply:
That would be incorrect. If `misa.B` is 1 then it implies that `Zba`, `Zbb`, and `Zbs` are present. However, if `misa.B` is 0, it does not imply that `Zba`, `Zbb`, or `Zbs` are not present.

Reply:
I have removed the haveBExt() - it can be added back if needed later.

Reply:
Ah good point, I see what you mean. So would it make sense to do this?

```
misa[B] = bool_to_bits(haveZba() & haveZbb() & haveZbs());
```

That would ensure the bit is set correctly if someone edits `haveZb?()` to return `false`.

Reply:
That would still have the issue that having `Zba`, `Zbb`, and `Zbs` does not imply `B`.  It  is valid for `B` to be 0 and  have *all* or some of `Zba`, `Zbb`, and `Zbs`. I am not sure we  need to guard againt people modifying code  in wrong ways - they can do that in many other ways. When we have ISA string parsing then one  would be able to specify "B" in ISA string in which case they do not need to explicitly add `Zba`, `Zbb`, and `Zbs` to the ISA string and `misa.B` becomes 1. Or they can choose to provide any combination of `Zba`, `Zbb`, and `Zbs`  and `misa.B`  stays  0. Or they can choose to provide `B`  and combination of `Zba`, `Zbb`, and `Zbs`  in which case the ones not  specified are forced to 1 because of `B` and `misa.B` is 1.

Reply:
> I am not sure we need to guard againt people modifying code in wrong ways - they can do that in many other ways.

Fully disagree here. The way this bit works is clearly quite confusing and I don't think it's unlikely that people will change these hard-coded `haveExt()` functions. Anyone using the model for verification will have to do it (we do). And when they do it would be nice for it to prevent B being set incorrectly.

I'm happy for it to be an assertion instead if you like? `assert(misa[B] == 0b0 | have.. & have... & have...)`. Maybe that's better?

Do we need to handle writing this bit for writable misa?

Reply:
> I'm happy for it to be an assertion instead if you like? assert(misa[B] == 0b0 | have.. & have... & have...). Maybe that's better?

Yeah an  assertion could be added. I can  see need to for many such we would add to adress people modifying code in  ways that does   not  make sense.

> Do we need to handle writing this bit for writable misa?

When misa is writeable this bit can be written. Clearing  `misa.B`  does not  imply `Zba`, `Zbb`, or `Zbs` go away.

Reply:
I made an update that should make it more clearer and also guard against people adding bugs when changing code.

Reply:
Yeah this looks much better, thanks!

> When misa is writeable this bit can be written.

Even if the extensions are not supported? I would have thought that makes it read-only zero.

To be fair I think all of this code needs some clean up, e.g. it doesn't look like the A bit isn't handled properly at all either, so I think this is good enough to merge now.

Reply:
Actually, re-reading the code I now understand how the `have...` and `sys...` functions are supposed to work, so I think the fully correct thing is:

```
// Indicates whether the B extension is *supported* (not enabled).
// TODO: Add a CLI parameter.
function sys_enable_b() -> bool = true

...

misa[B] = bool_to_bits(sys_enable_b()); /* Bit-manipulation */

...

function haveBExt() -> bool = misa[B] == 0b1
// These sub-extensions can be enabled even if misa[B] is 0, and currently we do that.
// TODO: Make this configurable.
function haveZba()  -> bool = true | haveBExt()
function haveZbb()  -> bool = true | haveBExt()
function haveZbs()  -> bool = true | haveBExt()
```

I think the naming and organisation of these functions is a bit misleading so I made an issue to fix it: #495

Sorry for all the changes!

Reply:
updated.